service/history/historyEngine.go

```diff
@@ -645,14 +647,26 @@ func (e *historyEngineImpl) startWorkflowHelper(
 
 	workflowID := request.GetWorkflowId()
 	domainID := domainEntry.GetInfo().ID
+
 	// grab the current context as a lock, nothing more
+	// use a smaller context timeout to get the lock
+	getLockContext := ctx
+	var cancel context.CancelFunc
+	if deadline, ok := ctx.Deadline(); ok {
+		now := e.shard.GetTimeSource().Now()
+		ctxTimeout := now.Sub(deadline) / 2
```

Comment:
nit: will / 2 be too much? Probably leave around 1s left should be enough? 

May be create/reuse a util function similar to: https://github.com/uber/cadence/blob/master/service/matching/taskListManager.go#L508?

Reply:
+1

Reply:
Seems to me its safe to have a very short time out here. The worst case is we return a reasonable error to the user that the workflow is currently locked and they retry. I am thinking like 100ms is fine?

Reply:
Both calculate from the input context or use a fix timeout work here. The only concern with a fix timeout is, the fix timeout could be larger than the input context timeout. (If a customer define a context timeout as 10ms but we use 100ms for this context)

Reply:
Did I read the code correctly? Shouldn't it be deadline.Sub(now)?

Reply:
Good catch. Added a test for it